File: bindings/erlang/src/ut0311.erl
-module(ut0311).

-export([broadcast/2, send/3, listen/2]).

-record(config, {bind, broadcast, listen, debug}).

-define(CONNECT_TIMEOUT, 5000).
-define(READ_TIMEOUT, 2500).
-define(LOG_TAG, "ut0311").

broadcast(Config, Request) ->
    dump(Request, Config#config.debug),

    Addr = Config#config.broadcast,

    case gen_udp:open(0, [inet, binary, {active, true}, {broadcast, true}]) of
        {ok, Socket} ->
            Result = broadcast(Socket, Addr, Request, Config#config.debug),
            gen_udp:close(Socket),
            Result;
        {error, Reason} ->
            {error, Reason}
    end.

broadcast(Socket, DestAddr, Request, Debug) ->
    case sendto(udp, Socket, DestAddr, Request) of
        ok ->
            erlang:send_after(?READ_TIMEOUT, self(), timeout),
            read_all(Debug);
        {error, Reason} ->
            {error, Reason}
    end.


send(Config, {controller,_,"",_}, Request) ->
    udp_broadcast_to(Config, Request);

send(Config, {controller,_, Address, "tcp"}, Request) ->
    case resolve(Address) of
        {ok, Addr} ->
            tcp_send_to(Config, Addr, Request);
        {error, Reason} ->
            {error, Reason}
    end;

send(Config, {controller,_, Address, _}, Request) ->
    case resolve(Address) of
        {ok, Addr} ->
            udp_send_to(Config, Addr, Request);
        {error, Reason} ->
            {error, Reason}
    end.


udp_broadcast_to(Config, Request) ->
    dump(Request, Config#config.debug),

    Addr = Config#config.broadcast,

    case gen_udp:open(0, [inet, binary, {active, true}, {broadcast, true}]) of
        {ok, Socket} ->
            Result = send(udp,Socket, Addr, Request, Config#config.debug),
            gen_udp:close(Socket),
            Result;
        {error, Reason} ->
            {error, Reason}
    end.

udp_send_to(Config, Addr, Request) ->
    dump(Request, Config#config.debug),

    case gen_udp:open(0, [inet, binary, {active, true}]) of
        {ok, Socket} ->
            Result = send(udp, Socket, Addr, Request, Config#config.debug),
            gen_udp:close(Socket),
            Result;
        {error, Reason} ->
            {error, Reason}
    end.

tcp_send_to(Config,{Addr,Port}, Request) ->
    dump(Request, Config#config.debug),

    case gen_tcp:connect(Addr, Port, [inet, binary, {active, true}], ?CONNECT_TIMEOUT) of
        {ok, Socket} ->
            Result = send(tcp,Socket, Addr, Request, Config#config.debug),
            gen_tcp:close(Socket),
            Result;
        {error, Reason} ->
            {error, Reason}
    end.


% set-ip doesn't return a reply
send(Transport, Socket, DestAddr, <<16#17, 16#96, R/binary>>, _Debug) ->
    case sendto(Transport, Socket, DestAddr, <<16#17, 16#96, R/binary>>) of
        ok ->
            erlang:send_after(?READ_TIMEOUT, self(), timeout),
            {ok, none};
        {error, Reason} ->
            {error, Reason}
    end;

send(Transport, Socket, DestAddr, Request, Debug) ->
    case sendto(Transport, Socket, DestAddr, Request) of
        ok ->
            erlang:send_after(?READ_TIMEOUT, self(), timeout),
            read(Transport, Debug);
        {error, Reason} ->
            {error, Reason}
    end.


listen(Config, Handler) ->
    {Addr, Port} = Config#config.listen,

    case gen_udp:open(Port, [inet, binary, {active, false}, {ip, Addr}]) of
        {ok, Socket} ->
            spawn(fun() ->
                listen(Socket, Handler, Config#config.debug),
                gen_udp:close(Socket),
                Handler ! closed
            end),
            {ok, fun() -> gen_udp:close(Socket) end};
        {error, Reason} ->
            {error, Reason}
    end.


listen(Socket, Handler, Debug) ->
    case gen_udp:recv(Socket, 64) of
        {ok, {_, _, Packet}} ->
            dump(Packet, Debug),
            Handler ! {ok, Packet},
            listen(Socket, Handler, Debug);
        {error, closed} ->
            log:infof(?LOG_TAG, closed),
            {error, closed};
        {error, Reason} ->
            log:infof(?LOG_TAG, io_lib:format("error ~p", [Reason])),
            {error, Reason}
    end.

sendto(tcp, Socket, _, Request) ->
    case inet:setopts(Socket, [{send_timeout, 1000}]) of
        ok ->
            gen_tcp:send(Socket, Request);
        {error, Reason} ->
            {error, Reason}
    end;

sendto(_, Socket, DestAddr, Request) ->
    case inet:setopts(Socket, [{send_timeout, 1000}]) of
        ok ->
            gen_udp:send(Socket, DestAddr, Request);
        {error, Reason} ->
            {error, Reason}
    end.


read_all(Debug) ->
    read_all([], Debug).

read_all(Received, Debug) ->
    receive
        {udp, _, _, _, Packet} ->
            dump(Packet, Debug),
            read_all([Packet | Received], Debug);
        timeout ->
            {ok, Received}
    end.


read(tcp, Debug) ->
    receive
        {tcp, _, Packet} ->
            dump(Packet, Debug),
            {ok, Packet};
        timeout ->
            {error, timeout}
    end;

read(_, Debug) ->
    receive
        {udp, _, _, _, Packet} ->
            dump(Packet, Debug),
            {ok, Packet};
        timeout ->
            {error, timeout}
    end.

resolve(S) ->
    case re:run(S, "([0-9]+[.][0-9]+[.][0-9]+[.][0-9]+)(?::([0-9]+))?", [{capture, all_but_first, list}]) of
        {match, [Address]} ->
            {ok, Addr} = inet:parse_ipv4strict_address(Address),
            {ok, {Addr, 60000}};
        {match, [Address, Port]} ->
            {ok, Addr} = inet:parse_address(Address),
            {ok, {Addr, list_to_integer(Port)}};
        _ ->
            inet:parse_ipv4strict_address(S)
    end.


dump(Packet, true) ->
    bin2hex(Packet, 0);
dump(_, _) ->
    ok.


bin2hex(<<P:8/binary, Q:8/binary, Rest/binary>>, Offset) ->
    <<P1:8, P2:8, P3:8, P4:8, P5:8, P6:8, P7:8, P8:8>> = P,
    <<Q1:8, Q2:8, Q3:8, Q4:8, Q5:8, Q6:8, Q7:8, Q8:8>> = Q,

    F = "~2.16.0B ~2.16.0B ~2.16.0B ~2.16.0B ~2.16.0B ~2.16.0B ~2.16.0B ~2.16.0B",

    io:fwrite("   ~8.16.0B  ~s  ~s~n", [
        Offset,
        io_lib:format(F, [P1, P2, P3, P4, P5, P6, P7, P8]),
        io_lib:format(F, [Q1, Q2, Q3, Q4, Q5, Q6, Q7, Q8])
    ]),

    bin2hex(Rest, Offset + 16);

bin2hex(_, _) ->
    io:fwrite("~n"),
    ok.

File: bindings/lua/src/ut0311.lua
local ut0311 = {}
local socket = require("socket")
local READ <const> = 5 -- seconds
local READALL <const> = 2.5 -- seconds

local debug = false
local bind_address = "*"
local bind_port = 0
local broadcast_address = "255.255.255.255"
local broadcast_port = 60000
local listen_address = "*"
local listen_port = 60001

function ut0311.set_bind_address(addr)
    if not addr or addr == "" then
        bind_address = "*"
        bind_port = 0
    else
        local address, port = addr:match("^(.-):(%d*)$")

        if address and port and address ~= "" and port ~= "" then
            bind_address = address
            bind_port = tonumber(port)
        elseif address and port and port ~= "" then
            bind_address = "*"
            bind_port = tonumber(port)
        else
            bind_address = addr
            bind_port = 0
        end
    end
end

function ut0311.set_broadcast_address(addr)
    if not addr or addr == "" then
        broadcast_address = "255.255.255.255"
        broadcast_port = 60000
    else
        local address, port = addr:match("^(.-):(%d*)$")

        if address and port and address ~= "" and port ~= "" then
            broadcast_address = address
            broadcast_port = tonumber(port)
        elseif address and port and port ~= "" then
            broadcast_address = "255.255.255.255"
            broadcast_port = tonumber(port)
        else
            broadcast_address = addr
            broadcast_port = 60000
        end
    end
end

function ut0311.set_listen_address(addr)
    if not addr or addr == "" then
        listen_address = "*"
        listen_port = 60001
    else
        local address, port = addr:match("^(.-):(%d*)$")

        if address and port and address ~= "" and port ~= "" then
            listen_address = address
            listen_port = tonumber(port)
        elseif address and port and port ~= "" then
            listen_address = "*"
            listen_port = tonumber(port)
        else
            listen_address = addr
            listen_port = 60001
        end
    end
end

function ut0311.set_debug(enabled)
    debug = enabled
end

function ut0311.broadcast(request)
    local sock = socket.udp4()
    local ok, result = pcall(function()
        return broadcast(sock, request)
    end)

    sock:close()

    if not ok then
        error(result)
    else
        return result
    end
end

function ut0311.send(controller, request)
    if controller['address'] ~= "" and controller['transport'] == "tcp" then
        return tcp_sendto(controller['address'], request)
    end
    
    if controller['address'] ~= "" then
        return udp_sendto(controller['address'], request)
    end
       
    return udp_broadcast_to(request)
end

function ut0311.listen(handler)
    local f = function(packet)
        dump(packet)
        handler(packet)
    end

    local sock = socket.udp4()
    local ok, result = pcall(function()
        return listen_to(sock, f)
    end)

    sock:close()

    if not ok then
        error(result)
    else
        return result
    end
end

function broadcast(sock, request)
    dump(request)

    if sock:setsockname(bind_address, bind_port) ~= 1 then
        error("error binding to address")
    end

    sock:settimeout(READ, "b")
    sock:settimeout(READALL, "t")

    if sock:setoption("broadcast", true) ~= 1 then
        error("error setting SO_BROADCAST")
    end

    sock:sendto(request, broadcast_address, broadcast_port)

    return read_all(sock)
end

function udp_broadcast_to(request)
    local sock = socket.udp4()
    local ok, result = pcall(function()
        return send(sock, request)
    end)

    sock:close()

    if not ok then
        error(result)
    else
        return result
    end
end

function udp_sendto(address, request)
    dump(request)

    addr,port = addrport(address)

    local sock = socket.udp4()
    local ok, result = pcall(function()
        if sock:setsockname(bind_address, 0) ~= 1 then
            error("error binding to address")
        end

        sock:settimeout(READ, "b")
        sock:settimeout(READALL, "t")
        sock:setpeername('192.168.1.100',60000)
        sock:send(request)

        -- set-ip doesn't return a reply
        if string.byte(request, 2) == 0x96 then
            return {}, nil
        end

        local packet = sock:receive(1024)
        if packet and #packet == 64 then
            dump(packet)
            return packet
        end

        return nil
    end)

    sock:close()

    if not ok then
        error(result)
    else
        return result
    end
end

function tcp_sendto(address, request)
    dump(request)

    addr,port = addrport(address)

    local sock = socket.tcp4()
    local ok, result = pcall(function()
        if sock:connect(addr, port) ~= 1 then
            error("error connecting to ", addr, port)
        end

        sock:settimeout(READ, "b")
        sock:settimeout(READALL, "t")
        sock:send(request)

        -- set-ip doesn't return a reply
        if string.byte(request, 2) == 0x96 then
            return {}, nil
        end

        local packet = sock:receive('*a')
        if packet and #packet == 64 then
            dump(packet)
            return packet
        end


        return nil
    end)

    sock:close()

    if not ok then
        error(result)
    else
        return result
    end
end

function send(sock, request)
    dump(request)

    if sock:setsockname(bind_address, 0) ~= 1 then
        error("error binding to address")
    end

    sock:settimeout(READ, "b")
    sock:settimeout(READALL, "t")

    if sock:setoption("broadcast", true) ~= 1 then
        error("error setting SO_BROADCAST")
    end

    sock:sendto(request, broadcast_address, broadcast_port)

    -- set-ip doesn't return a reply
    if string.byte(request, 2) == 0x96 then
        return {}, nil
    end

    return read(sock)
end

function read_all(sock)
    local replies = {}

    while true do
        local packet = sock:receive(1024)
        if not packet then
            break
        elseif #packet == 64 then
            dump(packet)
            table.insert(replies, packet)
        end
    end

    return replies
end

function read(sock)
    while true do
        local packet = sock:receive(1024)
        if not packet then
            break
        elseif #packet == 64 then
            dump(packet)
            return packet
        end
    end

    return nil
end

function listen_to(sock, f)
    if sock:setsockname(listen_address, listen_port) ~= 1 then
        error("error binding to listen address")
    end

    sock:settimeout(0, "b")
    sock:settimeout(nil, "t")

    local recv = function(sock)
        local packet, status = sock:receive(1024)
        if status == "timeout" then
            coroutine.yield(sock)
        end
        return packet, status
    end

    local co = coroutine.create(function ()
        while true do
            local packet,status = recv(sock)
            if status == "closed" then
                break
            elseif packet and #packet == 64 then
                f(packet)
            end
        end
    end)

    while true do
        local status, res = coroutine.resume(co)
        if not res then
            print("AWOOGAH")
            break
        end
        socket.select({sock})
    end
end

function addrport(address)
    if address and address ~= "" then
        local addr, port = address:match("^(.-):(%d*)$")
        if addr and port and addr ~= "" and port ~= "" then
            return addr, tonumber(port)
        end

        return address, 60000
    end

    return "255.255.255.255", 60000
end

function dump(packet)
    if debug then
        local ix = 1
        for i = 0, 3 do
            io.write(string.format("   %08x ", i * 16))
            for j = 1, 8 do
                io.write(string.format(" %02x", string.byte(packet, ix)))
                ix = ix + 1
            end
            io.write(" ")
            for j = 1, 8 do
                io.write(string.format(" %02x", string.byte(packet, ix)))
                ix = ix + 1
            end
            io.write("\n")
        end
        io.write("\n")
    end
end

return ut0311
